### DISCUSSION TITLE
examples of sqlg in the wild

### DISCUSSION THREAD

**Discussion by rhubarb** [pietermartin/sqlg#459] [Q&A]
Are there some examples of sqlg in the wild?
Does it generally run behind gremlin-server or in some other configuration?
I'm thinking of using it for a project with the possiblity of swapping out for neptune or neo4j in the future, so relying on tinkerpop to abstract my dependencies... does that seem reasonable?

**Comment by pietermartin** [comment]
Hi,

I recommend running `Sqlg` embedded in your server. On `postgresql` `Sqlg` supports a distributed mode where you can run it embedded in multiple jvms. It uses postgres's notify mechanism to keep its meta data synchronised. 

Most of `Sqlg`s custom features are not available via `gremlin-server`. The schema creation api is not available via `gremlin-server`, so no indexes / partitions / custom identifiers... The various bulk modes are also not possible via `gremlin-server`.

`gremlin-server` is only required if you are planning on using a non jvm language.

The production systems I run `Sqlg` on, are very large and its working well. `Sqlg` manages databases with 100s of schemas and thousands of tables each with 100s of columns. Last I checked the total amount of tables was about 50 000 and total columns over 1 000 000. The db sizes are generally around 5TB or so.

However if you plan to swap out later for another provider then using any of the custom features are risky. They all have their own custom features also, so not sure if its possible to avoid.

**Reply by rhubarb** [reply]
Thanks for the quick answer.
The main reasons I'm interested in gremlin-server are
1) because I want to put a REST API in front of my graph database for querying... Is there another way of doing that?
2) I figured that would help with the swap-it-out-for-neo4j-later options... but perhaps if I stick to gremlin, using the server doesn't make a difference there?

On the original question of 'out in the wild' can you point to any uses of this in other products or applications on GitHub or elsewhere?

thanks again

**Comment by pietermartin** [comment]
"REST API in front of my graph database for querying".
In this case yes `gremlin-server` will be best. 

I do not design applications like that, nor is it the intended use of `Sqlg`. I view `Sqlg` and `gremlin` equivalent to JDBC. No ones wraps JDBC inside a rest server. Ideally REST calls are at the application level and at a much higher level of abstraction.

That said `Sqlg` implements Tinkerpop exactly so there is no problem using `gremlin-server`.
"but perhaps if I stick to gremlin, using the server doesn't make a difference there?" yes,

Regarding out in the wild there was a US medical company using it in production. I have not heard from them for some time so don't know what is happening there.
Apart from stars on github I have no further insight to `Sqlg` out there in the wild.
